0000-access-keys.md

```diff
@@ -0,0 +1,286 @@
+  /// When used, both account balance and the allowance is decreased.
+  pub allowance: Balance(u128),
+
+  // Permissions (WORK IN PROGRESS)
+  
+  /// The AccountID of the receiver of the transaction. The access key will restrict transactions to
+  /// only this receiver, unless full_access is set to true.
+  pub receiver_id: AccountId(String),
+  
+  /// If not None, the access key would be restricted to call only the given method name.
+  /// Unless the full_access is set to true.
+  pub method_name: Option<String>,
+  
+  /// Turns this access key into a full permissions public key, that is similar to the account-level public keys.
+  pub full_access: bool,
+}
+
+/// NOTE: This change removes account-level nonce and public keys.
+/// Key is `account_id`
+struct Account {
+  pub balance: Balance(u128),
+  pub code_hash: Hash,
+  /// Storage usage accounts for all access keys      
+  pub storage_usage: StorageUsage(u64),
+    /// Last block index at which the storage was paid for.
+  pub storage_paid_at: BlockIndex(u64),
+}
+```
+
+### Examples
+
+#### AccessKey as account-level public key
+
+If an AccessKey has the full access to the account and the allowance set to be the max value for u128, then
+it essentially acts as an account-level public key. Which means we can remove account-level
+public keys from the account struct and rely only on access keys.
+
+An access key example from user `vasya.near` with full access (using TBD `full_access` permission field):
+```rust
+/// vasya.near,a123bca2
+AccessKey {
+    nonce: 0,
+    
+    allowance: 340282366920938463463374607431768211455,
+    
+    receiver_id: "vasya.near",
+    
+    method_name: None,
+    
+    full_access: true,
+}
+```
+
+
+#### AccessKey for a dApp by a user
+
+This is a simple example where a user wants to use some dApp. The user has to authorize this dApp within their wallet, so the dApp knows who the user is, and also can issue simple function call transactions on behalf of this user.
+
+To create such AccessKey a dApp generates a new key pair and passes the new public key to the user's wallet in a URL. Then the wallet asks the user to create a new AccessKey with
+that points to the dApp. User has to explicitly confirm this in the wallet for AccessKey to be created.
+The new access key is restricted to be only used for the app’s contract_id, but is not restricted for any method name.
+The user also selects the allowance to some reasonable amount, enough for the application to issue transactions. The application might hint the user about this allowance.
+Now the app can issue function call transactions on behalf of the user’s account towards the app’s contract without requiring the user to sign each transaction.
+
+An access key example for chess app from user `vasya.near` (using old permissions model):
+```rust
+/// vasya.near,c5d312f3
+AccessKey {
+    nonce: 0,
+    
+    allowance: 1_000_000_000,
```

Comment:
Why is this allowance not 0?

Reply:
added a comment

Reply:
Allowance has to be non-zero for the transaction to even be considered by the runtime.